File: src/main/java/com/github/michael72/pumlsrv/AppParams.java
package com.github.michael72.pumlsrv;

import java.io.File;
import java.util.prefs.Preferences;

public class AppParams {
  public enum OutputMode {
    Default, Dark, Light
  }

  private int port;
  int offset;
  File includeFile;
  final boolean reload;
  OutputMode outputMode;
  boolean isMonoChrome;
  final boolean showBrowser;
  final boolean noStore;
  boolean checkForUpdates;

  public AppParams(final int port, final int offset, final File includeFile, final boolean reload,
      final OutputMode outputMode, final boolean isMonoChrome, final boolean showBrowser, final boolean noStore,
      final boolean checkForUpdates) {
    this.port = port;
    this.offset = offset;
    this.includeFile = includeFile;
    this.reload = reload;
    this.outputMode = outputMode;
    this.isMonoChrome = isMonoChrome;
    this.showBrowser = showBrowser;
    this.noStore = noStore;
    this.checkForUpdates = checkForUpdates;
  }

  public AppParams next() {
    ++offset;
    return this;
  }

  public AppParams same() {
    --port;
    ++offset;
    return this;
  }

  public int port() {
    return this.port + this.offset;
  }

  public void setPort(int newPort) {
    this.port = newPort;
    this.offset = 0;
  }

  public void swapDarkMode() {
    this.outputMode = (this.outputMode != OutputMode.Dark) ? OutputMode.Dark : OutputMode.Default;
  }

  public void swapLightMode() {
    this.outputMode = (this.outputMode != OutputMode.Light) ? OutputMode.Light : OutputMode.Default;
  }

  public void setDefaultMode() {
    this.outputMode = OutputMode.Default;
  }

  public void store() {
    if (this.noStore) {
      return;
    }
    Preferences prefs = Preferences.userNodeForPackage(AppParams.class);
    prefs.putInt("port", port());
    prefs.put("include", includeFile != null ? includeFile.getAbsolutePath() : "");
    prefs.put("outputMode", outputMode.toString());
    prefs.putBoolean("isMonoChrome", isMonoChrome);
    prefs.putBoolean("checkForUpdates", checkForUpdates);
  }

  public void load() {
    if (this.noStore) {
      return;
    }
    Preferences prefs = Preferences.userNodeForPackage(AppParams.class);
    this.port = prefs.getInt("port", port());
    String inc = prefs.get("include", "");
    this.includeFile = (inc.length() > 0) ? new File(inc) : null;
    this.outputMode = OutputMode.valueOf(prefs.get("outputMode", outputMode.toString()));
    this.isMonoChrome = prefs.getBoolean("isMonoChrome", isMonoChrome);
    this.checkForUpdates = prefs.getBoolean("checkForUpdates", checkForUpdates);
  }

  public void loadPort() {
    Preferences prefs = Preferences.userNodeForPackage(AppParams.class);
    this.port = prefs.getInt("port", port());
  }
}

File: src/main/kotlin/com/github/michael72/pumlsrv/AppParams.kt
package com.github.michael72.pumlsrv

import java.io.File
import java.util.prefs.Preferences

data class AppParams(
    var portStart: Int,
    var offset: Int = 0,
    val includeFile: File? = null,
    val reload: Boolean = false,
    var outputMode: OutputMode = OutputMode.Default,
    var isMonoChrome: Boolean = false,
    var showBrowser: Boolean = true,
    val noStore: Boolean = false,
    var checkForUpdates: Boolean = true,
    val loadDynamicJar: Boolean = true
) {
    enum class OutputMode {
        Default, Dark, Light
    }

    fun next(): AppParams {
        offset++
        return this
    }

    fun same(): AppParams {
        portStart--
        offset++
        return this
    }

    fun port(): Int = portStart + offset

    fun setPort(newPort: Int) {
        portStart = newPort
        offset = 0
    }

    fun swapDarkMode() {
        outputMode = if (outputMode != OutputMode.Dark) OutputMode.Dark else OutputMode.Default
    }

    fun swapLightMode() {
        outputMode = if (outputMode != OutputMode.Light) OutputMode.Light else OutputMode.Default
    }

    fun setDefaultMode() {
        outputMode = OutputMode.Default
    }

    fun store() {
        if (noStore) return
        
        val prefs = Preferences.userNodeForPackage(AppParams::class.java)
        prefs.putInt("port", port())
        prefs.put("include", includeFile?.absolutePath ?: "")
        prefs.put("outputMode", outputMode.toString())
        prefs.putBoolean("isMonoChrome", isMonoChrome)
        prefs.putBoolean("checkForUpdates", checkForUpdates)
        prefs.putBoolean("showBrowser", showBrowser)
    }

    fun load() {
        if (noStore) return
        
        val prefs = Preferences.userNodeForPackage(AppParams::class.java)
        portStart = prefs.getInt("port", port())
        outputMode = OutputMode.valueOf(prefs.get("outputMode", outputMode.toString()))
        isMonoChrome = prefs.getBoolean("isMonoChrome", isMonoChrome)
        checkForUpdates = prefs.getBoolean("checkForUpdates", checkForUpdates)
        showBrowser = prefs.getBoolean("showBrowser", showBrowser)
    }

    fun loadPort() {
        val prefs = Preferences.userNodeForPackage(AppParams::class.java)
        portStart = prefs.getInt("port", port())
    }
}